dist/tools/ethos/start_network.sh

```diff
@@ -18,33 +26,69 @@ remove_tap() {
 
 cleanup() {
     echo "Cleaning up..."
-    remove_tap
     ip a d fd00:dead:beef::1/128 dev lo
-    kill ${UHCPD_PID}
+    kill -0 "${UHCPD_PID}" >/dev/null 2>&1 && kill -TERM "${UHCPD_PID}"
+    kill -0 "${ETHOS_PID}" >/dev/null 2>&1 && kill -TERM "${ETHOS_PID}"
+    remove_tap
     trap "" INT QUIT TERM EXIT
+    RUNNING=0
 }
 
 start_uhcpd() {
-    ${UHCPD} ${TAP} ${PREFIX} > /dev/null &
+    "${UHCPD}" "${TAP}" "${PREFIX}" > /dev/null &
     UHCPD_PID=$!
 }
 
-PORT=$1
-TAP=$2
-PREFIX=$3
-BAUDRATE=115200
-UHCPD="$(readlink -f "${ETHOS_DIR}/../uhcpd/bin")/uhcpd"
 
 [ -z "${PORT}" -o -z "${TAP}" -o -z "${PREFIX}" ] && {
     echo "usage: $0 <serial-port> <tap-device> <prefix> [baudrate]"
     exit 1
 }
 
-[ ! -z $4 ] && {
+[ -n "$4" ] && {
     BAUDRATE=$4
 }
 
+RUNNING=1
 trap "cleanup" INT QUIT TERM EXIT
 
-
-create_tap && start_uhcpd && "${ETHOS_DIR}/ethos" ${TAP} ${PORT} ${BAUDRATE}
+create_tap
+start_uhcpd
+if kill -0 "${UHCPD_PID}" >/dev/null 2>&1; then
+    echo "uhcpd running, pid ${UHCPD_PID}"
+else
+    echo "uhcpd failed to start" >&2
+    exit 1
+fi
+# Run ethos as a background job, but redirect the stdin of this script to the
+# ethos process (using <&0).
+# We need to do this to remember the ethos PID to be able to clean up if this
+# script is terminated from the outside (for example when using kill -TERM $PID),
+# or else ethos may remain running after the script has been terminated.
+# Observed when running start_network.sh from within another script and the other
+# script uses kill to tell this script to quit.
+"${ETHOS}" "${TAP}" "${PORT}" "${BAUDRATE}" <&0 &
```

Comment:
Apparently, this construct has some issues with Debian `/bin/sh` (dash) which seem to ignore the `<&0` and treat it as if it were `</dev/null`. Working on a solution...

Reply:
If it works on bash, why not change the script to a bash script? It wouldn't be the first script in the tree that uses bashisms.

Ignore this if there is an easy sh(dash) alternative. :)

Reply:
It does not work on `bash` for me (ubuntu 14.04), added `echo $SHELL` in the script which is bash in my case, and also replaced `#! /bin/sh` by `#! /bin/bash` without success

```
make BOARD=samr21-xpro term 
make -C ethos
make[2]: Nothing to be done for 'all'.
make -C uhcpd
make[2]: Nothing to be done for 'all'.
sudo sh /home/harter/work/git/RIOT/dist/tools/ethos/start_network.sh /dev/ttyACM0 tap0 2001:db8::/64
/bin/bash
net.ipv6.conf.tap0.forwarding = 1
net.ipv6.conf.tap0.accept_ra = 0
uhcpd running, pid 21972
ethos running, pid 21973
----> ethos: sending hello.
----> ethos: activating serial pass through.
EOF from stdin
----> ethos: hello reply received
```